Sites Management Page: Hide 'Hidden' sites by default

danielbachhuber opened the pull request:
Fixes #66380

## Proposed Changes

Hides 'Hidden' sites from the List and Tile view by default, and adds a message indicating such at the bottom:

<img width="1313" alt="image" src="https://user-images.githubusercontent.com/36432/184878599-d347aa14-720d-4a8d-84ff-180a116751bc.png">

'Hidden' sites appear as soon as you start searching. If you click on the link in the message, `?show-hidden=true` is appended to the URL and 'Hidden' sites will always appear (with or without searching).

## Testing Instructions

1. Switch to the branch and allow the application to rebuild.
2. Clear your `calypso_store` in LocalStorage. Calypso will need to re-request site excerpts to load the `visible` property.
3. Navigate to `/sites-dashboard`
4. Observe your 'Hidden' sites aren't displayed in the list.
5. Start searching for a 'Hidden' site.
6. Observe the site appear in the search results.
7. Clear out your search.
8. Navigate to the bottom of the page.
9. Observe the message about hidden sites.
10. Click on the link within the hidden sites message.
11. Observe all of your sites appear, regardless of whether they're hidden.

danielbachhuber committed:
Sites Management Page: Hide 'Hidden' sites by default

danielbachhuber committed:
Avoid unnecessary variable definition

danielbachhuber committed:
Update tests for 675500f2bcd3f17522b1a79dfbcd29aae9376707

danielbachhuber commented on client/sites-dashboard/components/sites-dashboard.tsx:
```diff
@@ -105,6 +126,28 @@ export function SitesDashboard( { queryParams: { search, status = 'all' } }: Sit
+								<HiddenSitesMessage>
+									{ translate(
+										'%(hiddenSitesCount)d more hidden site. {{a}}Change{{/a}}.{{br/}}Use search to access it.',
+										'%(hiddenSitesCount)d more hidden sites. {{a}}Change{{/a}}.{{br/}}Use search to access them.',
```
These strings are reused from `<SiteSelector>` but I don't think they're quite right...

zaguiini replied:
What are your concerns with it? Sounds fine to me 🤔 We could change the CTA to `Display them anyway` or something...

wojtekn replied:
I think it makes sense to rename "Change" CTA as here it performs different action than in the sidebar selector, where it points to a page that allows changing visibility settings.

Maybe "Show all" as this is already used in other places and translated?

p-jackson replied:
I prefer the original behaviour of the `Change` button from the site switcher tbh.

p-jackson replied:
I've changed the CTA to "Show all"

zaguiini commented on client/sites-dashboard/components/sites-dashboard.tsx:
```diff
@@ -59,8 +61,26 @@ const DashboardHeading = styled.h1`
 `;
 
-export function SitesDashboard( { queryParams: { search, status = 'all' } }: SitesDashboardProps ) {
+const HiddenSitesMessage = styled.span`
+	color: var( --color-text-subtle );
+	display: block;
+	font-size: 14px;
+	padding: 16px 16px 24px;
+
+	a {
+		text-decoration: underline;
+
+		&:hover {
+			text-decoration: none;
+		}
+	}
+`;
```
Let's try using CSS objects here as it provides a better developer experience:
- built-in autocompletion since we're working with TS objects;
- type checking out of the box since TS will yell at you if the property name (key) doesn't conform to the CSS types.

danielbachhuber replied:
> Let's try using CSS objects here as it provides a better developer experience:

I'd prefer to follow the existing pattern in the file for now.

zaguiini commented on client/sites-dashboard/components/sites-dashboard.tsx:
```diff
@@ -59,8 +61,26 @@ const DashboardHeading = styled.h1`
+		&:hover {
+			text-decoration: none;
+		}
+	}
```
Let's not nest CSS declarations. We're crafting components and defining how a child behaves based on the parent doesn't make much sense. Think of each styled component as a regular React component.

danielbachhuber replied:
Switched to a styled component with edb5359f3e

zaguiini commented on client/sites-dashboard/components/sites-dashboard.tsx:
```diff
@@ -105,6 +126,28 @@ export function SitesDashboard( { queryParams: { search, status = 'all' } }: Sit
 							{ displayMode === 'tile' && (
 								<SitesGrid isLoading={ isLoading } sites={ filteredSites } />
 							) }
+							{ selectedStatus.hiddenCount && (
```
Here we need to check if the `hiddenCount` is greater than zero. We need to do that because despite `0` being a falsy value, React interprets it as literal `0` and displays it:

![image](https://user-images.githubusercontent.com/26530524/183982826-022a52bb-359b-476e-8b87-bbd4fb46234b.png)

```suggestion
							{ selectedStatus.hiddenCount > 0 && (
```

danielbachhuber replied:
> Here we need to check if the `hiddenCount` is greater than zero. We need to do that because despite `0` being a falsy value, React interprets it as literal `0` and displays it:

Learned something new! And obvious in hindsight 😄 

Fixed with 4dbb235570500a3d831ca587672350fcf33a4d4e

zaguiini commented on packages/components/src/sites-table/use-sites-table-filtering.tsx:
```diff
@@ -43,10 +45,16 @@ export function useSitesTableFiltering(
 
 		for ( const status of statuses ) {
 			status.count = filteredByStatus[ status.name ].length;
+			if ( ! showHidden ) {
+				filteredByStatus[ status.name ] = [ ...filteredByStatus[ status.name ] ].filter(
+					( site ) => site.visible
+				);
+				status.hiddenCount = status.count - filteredByStatus[ status.name ].length;
+			}
```
This is a little bit hard to read but I can't think of a more readable version.

sejas replied:
I think we don't need to destruct the array: 
```diff
- filteredByStatus[ status.name ] = [ ...filteredByStatus[ status.name ] ].filter(
+ filteredByStatus[ status.name ] = filteredByStatus[ status.name ].filter(
```

p-jackson replied:
Yep, unlike `.sort()` (which is maybe where this was copied from), `.filter()` already returns a fresh array.

Perhaps it's ever so slightly more efficient to pass `showHidden` as an argument to `filterSites()` so that we do all the filtering in one pass? `O(n)` instead of `O(2n)`?

zaguiini reviewed (commented):
This is working. One thing I don't really like about it is once I click the CTA, the hidden sites are mixed with the visible ones, so it's impossible to distinguish the change. From a user's perspective, I'd like to see which sites were hidden.

danielbachhuber committed:
Prevent count from being output

danielbachhuber committed:
Use a styled component instead

danielbachhuber commented:
> One thing I don't really like about it is once I click the CTA, the hidden sites are mixed with the visible ones, so it's impossible to distinguish the change. From a user's perspective, I'd like to see which sites were hidden.

@zaguiini This is a good observation. It seems like we'd need some design direction on how to handle.

wojtekn commented on client/data/sites/site-excerpt-constants.ts:
```diff
@@ -5,6 +5,7 @@ export const SITE_EXCERPT_REQUEST_FIELDS = [
 	'URL',
 	'is_coming_soon',
 	'is_private',
+	'visible',
```
I try to understand why we need that - that data, including the `visible` key, is already available in the `sites` store. What is the use case in which we need this?

danielbachhuber replied:
> that data, including the `visible` key, is already available in the `sites` store. What is the use case in which we need this?

If the `sites` store isn't yet loaded, we use `useSiteExcerptsQuery` instead. This `SITE_EXCERPT_REQUEST_FIELDS` constant is used to determine the fields loaded by `useSiteExcerptsQuery`.

sejas commented:
@danielbachhuber is there a way to hide the sites again? I couldn't find it.

danielbachhuber commented:
> is there a way to hide the sites again? I couldn't find it.

@sejas At the moment, no.

p-jackson commented on client/sites-dashboard/components/sites-dashboard.tsx:
```diff
@@ -59,8 +61,26 @@ const DashboardHeading = styled.h1`
+	color: var( --color-text-subtle );
+	display: block;
+	font-size: 14px;
+	padding: 16px 16px 24px;
```
Personally I prefer 
```suggestion
	padding: 16px 0 24px 0;
```

Which aligns the left of the text with the left of the rows.

![CleanShot 2022-08-15 at 14 56 18@2x](https://user-images.githubusercontent.com/1500769/184569724-0e09162d-e74c-4a86-9ea5-e921fe176322.png)

p-jackson commented on client/sites-dashboard/components/sites-content-controls.tsx:
```diff
@@ -10,8 +10,9 @@ import { SitesSearch } from './sites-search';
 export interface SitesDashboardQueryParams {
-	status?: string;
 	search?: string;
+	showHidden?: boolean;
```
Do we even need it to be an option in the URL? I suppose it allows someone to type in a search query and set `show-hidden=false`, but who's gonna do that? I feel like setting `search` should implicitly mean to show hidden.

danielbachhuber replied:
Per conversation yesterday, the `?show-hidden=true` is added to the URL because it represents a filter state of the table.

p-jackson commented on client/sites-dashboard/components/sites-dashboard.tsx:
```diff
@@ -107,6 +128,28 @@ export function SitesDashboard( { queryParams: { search, status = 'all' } }: Sit
+												br: <br />,
+												a: (
+													<HiddenSitesMessageLink
+														href={ addQueryArgs( window.location.href, { 'show-hidden': 'true' } ) }
```
I think this should link to `https://dashboard.wordpress.com/wp-admin/index.php?page=my-blogs&show=visible`, just like the site switcher.

p-jackson commented on client/sites-dashboard/controller.tsx:
```diff
@@ -24,6 +24,7 @@ export function sitesDashboard( context: PageJSContext, next: () => void ) {
 			<SitesDashboard
 				queryParams={ {
 					search: context.query.search,
+					showHidden: context.query[ 'show-hidden' ],
```
At runtime the value of `showHidden` is going to end up being a string, even though the prop says its a `boolean`. Generally that's going to work out ok, but could lead to some surprising behaviour down the track where the TypeScript engine insists that something isn't possible, even though at rutime it is possible.

I mentioned above that I don't actually think we need the `show-hidden` query param, but if we are going to keep it maybe we should for this into a boolean and make sure `show-hidden=false` has the expected behaviour.
```suggestion
					showHidden: context.query[ 'show-hidden' ] === 'true',
```

p-jackson reviewed (commented):
Because the sites are hidden at a pretty global level, I wasn't expecting the site counts in the dropdown to include the hidden sites.
![CleanShot 2022-08-15 at 16 50 25@2x](https://user-images.githubusercontent.com/1500769/184578643-5959e5c5-6e2b-42c5-b18c-b7c413b70fc5.png)

But then on the other hand maybe it'd be a bit strange for the number in the dropdown to jump from ~10 to ~100 as soon as you start typing.

I also think that we could do with some more vertical padding when in grid mode.
![CleanShot 2022-08-15 at 16 54 12@2x](https://user-images.githubusercontent.com/1500769/184580380-2d336ce8-9df3-46e4-a13d-240ca959d7bf.png)

p-jackson committed:
Use a "Show all" button to unhide all sites

p-jackson committed:
Share padding at bottom of table and grid

p-jackson committed:
Ensure the `showHidden` prop is a boolean

Otherwise the type according to Typescript isn't going to match the
runtime type.

p-jackson committed:
Don't include hidden sites in the site counts displayed in dropdown

p-jackson commented:
I had a chat with @danielbachhuber about how I was in favour of following the same behaviour as the site switcher (i.e. no special filter for showing hidden sites). My reservations came down to (1) we seemed to be changing the level of "hidden-ness" that this flag represents (i.e. is it supposed to be very easy to see them, or a little bit inconvenient?) and (2) the UI was almost an exact match to the site switcher, but the way it works is a little different, and I thought was confusing.

But after talking it out I'm satisfied with the proposed behaviour.

For (1), the site visibility flag pre-dates Calypso, but back when it was introduced it would have been part of the dashboard interface that everyone used. So really we're reinstating the "hidden-ness" that was originally intended by this flag.

For (2) we discussed a different sort of UI, perhaps something that's reminiscent of an infinite scroll, that has a "show more" button at the bottom of a long list. Google Photos does this.
![CleanShot 2022-08-16 at 13 45 43@2x](https://user-images.githubusercontent.com/1500769/184794420-4095bdc3-9af7-4f75-94cc-2ffea5ac9ee7.png)


A couple of other changes I've made:
- Changed the counts in the status dropdown to exclude hidden sites when they're not visible. This might seem a little inconsistent given that sites which have been excluded by fuzzy search are still included in the count. But I think of the `?show-hidden=true` filter as a sort of uber page-wide filter. The idea is that most of the time you can just ignore those hidden sites exist. Same reason why we don't have a dropdown control for showing hidden sites.
- Ensured the table and grid have the same margin at the bottom. This is to make sure there's enough padding between the table/grid and footer message. This has made the CSS code for `<SitesDashboard>` a bit messy, requiring both `@emotion/css` and `@emotion/react` to be imported. I'll address this in a follow up PR

danielbachhuber committed:
Wrap 'Show All' in a translation function

danielbachhuber committed:
Tweak language to be more specific about what 'hidden' means

p-jackson committed:
Switch <SitesDashboard> to @emotion/css (#66588)

The component was generating some CSS class names using @emotion/react,
and some using @emotion/css. We don't need that. Just use @emotion/css
for everything.

Co-authored-by: Daniel Bachhuber <daniel.bachhuber@automattic.com>

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7476943

Thank you @danielbachhuber for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7476943).
